### ISSUE TITLE
Please support SameSite=None and Secure when setting cookies

### ISSUE THREAD

**Issue by davidmi** [Sustainsys/Saml2#1091]
Google announced that they will soon be requiring a new explicit SameSite=None property in order to send cookies on cross site requests. 

https://blog.chromium.org/2019/05/improving-privacy-and-security-on-web.html
https://web.dev/samesite-cookies-explained/

With Chrome 76, as described in the second link, with the `same-site-by-default-cookies` and  `cookies-without-same-site-must-be-secure` flags set, cookies will no longer be sent when the IdP POSTs to the ACS unless `SameSite=None; Secure` was set on the cookie, which will break the current implementation.

Please support setting this new header so that sites using Sustainsys do not break when this change arrives.

Related, I've noticed that in ASP.NET Core that setting SameSiteAttribute.None does not explicitly send a `SameSite: None` cookie attribute as that was assumed by the implementation and previous specification to be the default. I worked around this by manually setting the cookie with 

```
response.Headers.Append("set-cookie", $"{key}={value}; path=/; SameSite=None; Secure");
```

Similar workarounds will probably be necessary for other framework implementations.

### Related issues
This is the main issue for tracking the SameSite updates. There are also separate issues for each API package on the 1.X and 2.X versions. Please see list of related issues below.

**Comment by AndersAbel** [comment]
Thanks for bringing this up, looks like something that needs to be added ASAP.

**Comment by explunit** [comment]
The information here should help determine the timing: https://www.chromestatus.com/feature/5088147346030592

It looks like we can't make the change until Apple fixes their Safari bug.

**Comment by davidmi** [comment]
This is specific to how we use SustainSys.Saml2, but after executing SignInCommand.Run() directly we reimplement https://github.com/Sustainsys/Saml2/blob/83b53c3790386f0e39b1d33554b62b222b6cb249/Sustainsys.Saml2.AspNetCore2/CommandResultExtensions.cs#L29

                var cookieData = HttpRequestData.ConvertBinaryData(
                    dataProtector.Protect(commandResult.GetSerializedRequestState()));

                httpContext.Response.Cookies.Append(
                    commandResult.SetCookieName,
                    cookieData,
                    new CookieOptions()
                    {
                        HttpOnly = true,
                        // We are expecting a different site to POST back to us,
                        // so the ASP.Net Core default of Lax is not appropriate in this case
                        SameSite = SameSiteMode.None,
                        IsEssential = true
                    });

The issue is that SameSiteMode.None does not actuallly explicitly send SameSite=None as a header, so we do the same as above but instead of Response.Cookies.Append we do 
                 response.Headers.Append("set-cookie", $"{key}={value}; path=/; SameSite=None; Secure; HttpOnly")

The following is the issue in ASP.NET Core tracking the behavior of SameSiteMode=None
https://github.com/aspnet/AspNetCore/issues/12125 - they haven't changed it yet.

**Comment by duncansmart** [comment]
More information : https://devblogs.microsoft.com/aspnet/upcoming-samesite-cookie-changes-in-asp-net-and-asp-net-core/ 

TL;DR: there's a patch coming on November 19th for .NET 4.7.2 and .NET Core 2.1 and later to support setting `SameSite=None` but you need to use User-Agent sniffing so you don't send this to certain versions of iOS and macOS Safari 😬

**Comment by AndersAbel** [comment]
> More information : https://devblogs.microsoft.com/aspnet/upcoming-samesite-cookie-changes-in-asp-net-and-asp-net-core/
> 
> TL;DR: there's a patch coming on November 19th for .NET 4.7.2 and .NET Core 2.1 and later to support setting `SameSite=None` but you need to use User-Agent sniffing so you don't send this to certain versions of iOS and macOS Safari 😬

Thanks, I've seen that too. Sometimes after the patches land on November 19th I will release an updated version that supports this. I think that this will also move the minimum supported version of .NET Framework to 4.7.2. This will of course need to be done well before February when Chrome deploys their new version.

Fixing SameSite mode of the correlation cookie won't be enough though, there are definitely problems with single logout too. Processing an incoming Single Logout request requires access to the session cookie. And we definitely don't want to make that SameSite=None. Either single logout over POST needs to be removed completely, or we need to come up with some double post/refresh scheme that renders another internal POST page which is treated as samesite by all browsers.

**Comment by mklinke** [comment]
FYI, there is a related [breaking change in .NET Core 3.1](https://docs.microsoft.com/en-us/dotnet/core/compatibility/3.0-3.1), where the behavior of `SameSiteMode.None` is changed and a new value `SameSiteMode.Unspecified` is introduced.

**Comment by ulrichb** [comment]
Okay, ASP.NET now emits the `SameSite=None` value. But what about the `Secure` flag?

Chrome >= 80 (just tested) drops the correlation cookie:

![image](https://user-images.githubusercontent.com/388796/70761396-5e4ccc00-1d4d-11ea-948f-018267926adc.png)

=> Set the Secure flag to `HttpContext.Request.IsHttps` in [CommandResultExtensions](https://github.com/Sustainsys/Saml2/blob/2967605027bf825c17b895defe61c387fb2c7735/Sustainsys.Saml2.AspNetCore2/CommandResultExtensions.cs#L32-L42) ?

**Comment by AndersAbel** [comment]
> => Set the `Secure` prop to `HttpContext.Request.IsHttps` in [CommandResultExtensions](https://github.com/Sustainsys/Saml2/blob/2967605027bf825c17b895defe61c387fb2c7735/Sustainsys.Saml2.AspNetCore2/CommandResultExtensions.cs#L32-L42) ?

Sounds like a good idea. But it should probably also look at the scheme of `PublicOrigin` if that is used with an external TLS termination device.

If possible I would prefer to keep the possibility to run the library on plain http - it's a bad idea for production, but might be useful during development.

**Comment by wkrosner** [comment]
has this been resolved?

**Comment by lennart77** [comment]
Hi all,

We are a bit worried about the silence regarding this issue. It is our understanding that without this fix, POST-based SAML2 logon will break in Chrome with the rollout of Chrome 80. At least if the logon process on the IdP takes more than 2 minutes, which is not unlikely if the user is in a setting where they need to enter username, password and some two-way authentication token.

Roll out of Chrome 80 is happening in February, which is not so far away. Will this be addressed by then?

Here are some information:
https://www.chromium.org/updates/same-site
https://blog.chromium.org/2019/10/developers-get-ready-for-new.html

**Comment by Moosjon** [comment]
@AndersAbel In October 2019 you wrote "Sometimes after the patches land on November 19th I will release an updated version that supports this." and "This will of course need to be done **well before February** when Chrome deploys their new version.". 

Are you still planning on releasing updated versions anytime soon and is there any progress, or do we have to start looking for other solutions or fixes? It's almost halfway januari now and these issues are becoming urgent. We are using the .NET Framework HttpModule version, so I really hope the fix is coming both for .NET .NET Framework and .NET Core versions.

Thanks in advance for your reaction/time!

**Comment by AndersAbel** [comment]
There will be a release within a week from today. Patches will be released to 1.0.1 and 2.3.0 versions.

**Comment by lennart77** [comment]
That is great news @AndersAbel. Thanks!

**Comment by AndersAbel** [comment]
First fix (of 8) pushed. There are some design choices in here, especially on how to NOT send `SameSite=None` for non-compatible browsers. There is a new notification that is called to decide if `SameSite=None` will be sent. Feedback on that design is welcome.

**Comment by AndersAbel** [comment]
Another update - the target Framework for the MVC and HttpModule packages will be 4.7.2, as that's the lowest that can control the SameSite flag. I'll try to keep the Owin package to 4.5.2 (thanks @explunit for the catch that it should be possible).

**Comment by rikrak** [comment]
How does this change to Chrome manifest as an issue in a SAML2 sign in?  I've just been testing my application (that uses the Kentor libraries - v0.19.0) against ADFS over https using Chrome Beta (v80.0.3987.53), but I am able to login successfully.
I wondering if my login scenario is unaffected by this change to Chrome, or am I missing something?

**Comment by AndersAbel** [comment]
@rikrak If you are quick enough to do the login, it'll will work due to the two minutes grace period that Chrome has in version 80 (it will be removed later). The best way I've found out to see what's happening is to perform the log in while having the Javascript console open - some information is printed there.
If you're still on Kentor.AuthServices you should really change anyways, those libraries are not receiving security fixes. Only Sustainsys.Saml2 1.X and Sustainsys.Saml2 2.X (where X is latest available) will be patched.

**Comment by AndersAbel** [comment]
The to-be version 1.0.1 is now in the 1.0 branch. I've updated unit tests and run manual tests, but further help to test and validate the new behaviour is appreciated.

Tomorrow I'll port the patches to master for the 2.4.0 release.

**Comment by Moosjon** [comment]
@AndersAbel: Thanks for the updates!  
I tested my project locally using the updated 1.0 branch HttpModule in combination with your StupIdp, and it's working again. Firefox implemented the same flags for SameSite as Chrome (in about:config), but without the two minutes grace period. Before the changes the login wasn't working at all in Firefox.

**Comment by AndersAbel** [comment]
And patch merged to master. All unit tests run, but I've not yet tested the sample applications. Thought I'd better push it anyway to make it available for testing for other. Please report back if you find any problems!

**Comment by chrisnott** [comment]
Am I right in noticing that you have upgraded to Katana 4? Does this mean that the SameSite changes will not be available when integrating with IdentityServer3?

**Comment by AndersAbel** [comment]
@chrisnott Yes, you are right that I have upgraded to Katana 4 as that's the lowest version that supports setting the SameSite attributes. Is it possible to have your application reference IdSrv3 but still reference the Katana 4 packages in your hosting application? That way they could work together.

If you are using IdentityServer3, you also have correlation cookies in there that is not fixed in any public release. There is a patch available under the IdentityServer3 commercial support package.

**Comment by chrisnott** [comment]
Thanks @AndersAbel. I will have a go and see what I can achieve. I am aware of the IdentityServer3 security maintenance programme but thanks for that additional piece of information. Very useful.

**Comment by cip1337** [comment]
After updating to 2.4.0 most of the cookies are ok now beside saml2 one on return from Okta. We are using integration sustainsys - Okta because we have a local idp:

FF:
![image](https://user-images.githubusercontent.com/26924453/72714755-ea85b680-3b77-11ea-97f5-66591ac12319.png)

This has the secure flag but not SameSite=None flag.

Chrome also:
![image](https://user-images.githubusercontent.com/26924453/72714832-143edd80-3b78-11ea-99fd-60df2d145f99.png)

**Comment by AndersAbel** [comment]
> After updating to 2.4.0 most of the cookies are ok now beside saml2 one on return from Okta. We are using integration sustainsys - Okta because we have a local idp:

What package are you using? Is this Asp.Net Core? Owin?

**Comment by raxitr** [comment]
> @chrisnott Yes, you are right that I have upgraded to Katana 4 as that's the lowest version that supports setting the SameSite attributes. Is it possible to have your application reference IdSrv3 but still reference the Katana 4 packages in your hosting application? That way they could work together.
> 
> If you are using IdentityServer3, you also have correlation cookies in there that is not fixed in any public release. There is a patch available under the IdentityServer3 commercial support package.

We are on IdentityServer3 security maintenance program and learned IdentityServer3 is not compatible with Katana 4. This means we cannot upgrade to Sustainssys Saml2 1.0.1, what options do we have?

**Comment by AndersAbel** [comment]
@raxitr I think that you can run 1.0.0 and implement AcsCommandResultCreated, add your own cookie setting mechanism and set CommandResult.Handled to true to skip the built in cookie setting (you also need to reimplement/copy the rest of Sustainsys.Saml2.Owin.CommandResultExtensions.Apply.

If you want further info on how to do this, I would be happy to help under my commercial support offering, please mail me at anders@sustainsys.com. I guess some coordination is needed with the non-public IdentityServer3 maintenance release.

**Comment by raxitr** [comment]
Are you suggesting I do source modification in 1.0.0 and change the cookie setting mechanism to my own? 

Here is what I have done:
I upgraded my package to 1.0.0 and wrote new Sustainsys.Saml2.Owin.CommandResultExtensions.Apply method with my own cookie setting mechanism but I don't seem to find a way to call it.

**Comment by AndersAbel** [comment]
> Are you suggesting I do source modification in 1.0.0 and change the cookie setting mechanism to my own?
No

> I upgraded my package to 1.0.0 and wrote new Sustainsys.Saml2.Owin.CommandResultExtensions.Apply method with my own cookie setting mechanism but I don't seem to find a way to call it.

Hook up the `AcsCommandResultCreated` notification. It is called immediately before the Apply method. In your custom method, set the `HandledResult` flag to not run the built in `Apply` method. Unfortunately there is no way to only replace the cookie setting, you also do need to set result code, write body, set headers etc. That's why I suggested that you start by copying the code from the existing method.

**Comment by raxitr** [comment]
I will share my workaround which seem to have fixed it but that is not ideal as I have copied Owin files and tweaked extension method "Apply" so that it is aware of Samesite check.

I would like to implement your suggestion but I do not understand when to call Apply method. Since Apply method is an extension method I don't seem to find a way to call it because it expects OwinContext and DataProtector which is not available in AcsCommandResultCreated notification. Am I missing anything?

**Comment by jreames9** [comment]
has anyone gotten the AcsCommandResultCreated notification override to work under 1.0.0? I'm in the same boat as @raxitr so i've got to figure out how to get it to work with the patched ID3, but I'm not able to ever get it to trigger that notification. I've gotten AuthenticationRequestCreated and SignInCommandResultCreated to trigger but not AcsCommandResultCreated.

**Comment by raxitr** [comment]
@jreames9 I could not figure AcsCommandResultCreated but got it working with some source code modification in 1.0.0. Let me know if you want to go that route.

**Comment by jreames9** [comment]
@raxitr i did the same thing yesterday and so far that is working for me as well. thanks for the update.

**Comment by JasminSAB** [comment]
Hello, I have the exact problem with ASP.NET Core Blazor hosted application. ((Specify SameSite=None and Secure if the cookie is intended to be set in cross-site contexts. Note that only cookies sent over HTTPS may use the Secure attribute.) 

Application is deployed, and a problem occurs with the following cookies:

Identity.External 

Note that I'm using SSL yet on my app, but still this should work and be supported out of the box.
